dm/pkg/checker/worker_pool.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+// WorkerPool is a easy-to-use worker pool that can start workers by Go then use
+// PutJob to send jobs to worker. After worker finished a job, the result is
+// sequentially called by resultHandler function which is the parameter of
+// NewWorkerPool or NewWorkerPoolWithContext. After caller send all jobs, it can
+// call Wait to make sure all jobs are finished.
+// The type parameter J means job, R means result. Type J MUST only share
+// concurrent-safe member like *sql.DB.
+type WorkerPool[J, R any] struct {
+	ctx context.Context
+
+	// the closing order is inCh -> outCh -> done
+	inCh     chan J
+	outCh    chan R
+	done     chan struct{}
+	errGroup *errgroup.Group
+}
+
+// NewWorkerPool creates a new worker pool.
+// The type parameter J means job, R means result. Type J MUST only share
+// concurrent-safe member like *sql.DB.
+func NewWorkerPool[J, R any](resultHandler func(R)) *WorkerPool[J, R] {
+	return NewWorkerPoolWithContext[J, R](context.Background(), resultHandler)
+}
+
+// NewWorkerPoolWithContext creates a new worker pool with a context which may
+// be canceled from caller.
+// The type parameter J means job, R means result. Type J MUST only share
+// concurrent-safe member like *sql.DB.
+func NewWorkerPoolWithContext[J, R any](
+	ctx context.Context,
+	resultHandler func(R),
```

Comment:
- maybe add `concurrency` param here
- no need a public `Go`, loop and call `Go` is common pattern
- maybe let J impl this interface
```go
type JobI interface {
    Run(ctx) R
}
```

Reply:
> maybe add concurrency param here
> no need a public Go, loop and call Go is common pattern

seems now the worker pool will spawn workers, but if one worker need init logic for multiple jobs, like create a downstream parser, the init logic should also be passed to `NewWorkerPoolWithContext` which is hard to design.

Now we let the user go init logic in for-loop then call `Go`

> maybe let J impl this interface

it's also related to above thing. Choosing let worker owns `run` rather than job owns `run` can have worker init logic for multiple jobs

Reply:
> but if one worker need init logic for multiple jobs, , like create a downstream parser

why let worker-pool care about job init? let job do it or we init the parser for job before `PutJob`

Reply:
it's worker init logic, not job init. For example, there're 16 workers and each should get a downstream parser. The number of job is not determined, we only know the worker numer is 16.

Also job has its own init logic, such as upstream DB connection pool, but worker init is another thing